packages/shared/src/hooks/useUserSettingsLocalStorage.ts

```diff
@@ -0,0 +1,187 @@
+    const configMapData = deseralizeData(newValue);
+    const newData = configMapData?.[userSettingsKey];
+
+    if (newData !== undefined && seralizeData(newData) !== seralizeData(data)) {
+      setData(newData);
+    }
+  }
+};
+
+const updateLocalStorage = <T>(
+  storageKey: string,
+  userSettingsKey: string,
+  configMapData: any,
+  newState: T
+) => {
+  // Trigger update also when unmounted
+  const dataToUpdate = {
+    ...configMapData,
+    ...{
+      [userSettingsKey]: newState,
+    },
+  };
+
+  const newValue = seralizeData(dataToUpdate);
+
+  try {
+    // Update the local storage
+    localStorage.setItem(storageKey, newValue);
+
+    // Dispatch storage event to sync the settings changes for other tabs
+    generateStorageEventToUpdate(storageKey, newValue);
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.error(
+      `Error while updating local storage for key ${storageKey}`,
+      err
+    );
+  }
+};
+
+// Generate storage event to sync latest settings on other tabs
+const generateStorageEventToUpdate = (storageKey, newValue: string) => {
+  // Create a storage event to dispatch locally since browser windows do not fire the
+  // storage event if the change originated from the current window
+  const event = new StorageEvent('storage', {
+    storageArea: localStorage,
+    key: storageKey,
+    newValue,
+    oldValue: localStorage.getItem(storageKey),
+    url: window.location.toString(),
+  });
+
+  window.dispatchEvent(event);
+};
+
+export const useUserSettingsLocalStorage = <T>(
+  // User settings sub storage path
+  userSettingsKey: string,
+  // Sync user settings changes in all tabs
+  sync = false,
+  // Default value for the initial settings
+  defaultValue?: T,
+  // ODF settings base storage path
+  storageKey: string = DEFAULT_ODF_CONSOLE_USER_SETTINGS_KEY
+): [T, React.Dispatch<React.SetStateAction<T>>] => {
+  // Mount status for safty state updates
+  const mounted = React.useRef(true);
+  React.useEffect(() => () => (mounted.current = false), []);
+  const keyRef = React.useRef<string>(
+    userSettingsKey?.replace(/[^-._a-zA-Z0-9]/g, '_')
+  );
+  const defaultValueRef = React.useRef(defaultValue);
+
+  const [data, setData] = React.useState(() =>
+    intializeDataFromLocalStorage(
+      storageKey,
+      keyRef.current,
+      defaultValueRef.current
+    )
+  );
+  const dataRef = React.useRef<T>(data);
+  dataRef.current = data;
+
+  // Sync latest user settings across all tabs.
+  const syncLatestData = React.useCallback(
+    (event: StorageEvent) =>
+      syncLatestDataFromLocalStorage(
+        mounted.current,
+        storageKey,
+        event.storageArea,
+        event.key,
+        event.newValue,
+        dataRef.current,
+        keyRef.current,
+        setData
+      ),
+    [storageKey]
+  );
+
+  React.useEffect(() => {
+    if (sync) {
+      window.addEventListener('storage', syncLatestData);
```

Comment:
nits: const STORAGE_EVENT = 'storage';
```suggestion
      window.addEventListener(STORAGE_EVENT, syncLatestData);
```